src/RevSysCalls.cc

```diff
@@ -595,65 +598,55 @@ EcallStatus RevProc::ECALL_read(RevInst& inst){
   return EcallStatus::SUCCESS;
 }
 
-// 64, rev_write(unsigned int fd, const char  *buf, size_t count)
+
 EcallStatus RevProc::ECALL_write(RevInst& inst){
-  output->verbose(CALL_INFO, 2, 0, "ECALL: write called on Hart %" PRIu32 "by thread %" PRIu32 " on hart %" PRIu32 "\n", HartToExec, GetActiveThreadID(), HartToExec);
+  output->verbose(CALL_INFO, 2, 0, "ECALL: write called by thread %" PRIu32 " on hart %" PRIu32 "\n",  GetActiveThreadID(), HartToExec);
   auto fd = RegFile->GetX<int>(RevReg::a0);
   auto addr = RegFile->GetX<uint64_t>(RevReg::a1);
   auto nbytes = RegFile->GetX<uint64_t>(RevReg::a2);
-  auto rtv = EcallStatus::ERROR;
 
   auto& ECALL = Harts.at(HartToExec)->GetEcallState();
+  auto lsq_hash = make_lsq_hash(10, RevRegClass::RegGPR, HartToExec); // Cached hash value
 
-  // RevRegFile* RegFile = AssignedThreads.at(Harts.at(HartToExec)->GetAssignedThreadID())->GetRegFile();
-
-  if(ECALL.bytesRead){
-    // Not our first time through... so capture previous read data
+  if(ECALL.bytesRead && LSQueue->count(lsq_hash) == 0){
     ECALL.string += std::string_view(ECALL.buf.data(), ECALL.bytesRead);
     ECALL.bytesRead = 0;
   }
 
   auto nleft = nbytes - ECALL.string.size();
-  if(nleft == 0){
-    // Perform the write on the host system
+  if(nleft == 0 && LSQueue->count(lsq_hash) == 0){
     int rc = write(fd, ECALL.string.data(), ECALL.string.size());
-
-    // write returns the number of bytes written
     RegFile->SetX(RevReg::a0, rc);
-
-    // Reset our tracking state
     ECALL.clear();
-
     DependencyClear(HartToExec, 10, false);
-    rtv = EcallStatus::SUCCESS;
-  }else if (0 == LSQueue->count(make_lsq_hash(10, RevRegClass::RegGPR, HartToExec)))  {
-    auto readfunc = [&](auto* buf){
-      MemReq req (addr + ECALL.string.size(), 10, RevRegClass::RegGPR, HartToExec, MemOp::MemOpREAD, true, RegFile->GetMarkLoadComplete());
-      LSQueue->insert({make_lsq_hash(req.DestReg, req.RegType, req.Hart), req});
-      mem->ReadVal(HartToExec,
-                   addr + ECALL.string.size(),
-                   buf,
-                   req,
-                   REVMEM_FLAGS(0));
-      ECALL.bytesRead = sizeof(*buf);
-    };
+    return EcallStatus::SUCCESS;
+  }
+
+  if (LSQueue->count(lsq_hash) == 0) {
+    MemReq req (addr + ECALL.string.size(), 10, RevRegClass::RegGPR, HartToExec, MemOp::MemOpREAD, true, RegFile->GetMarkLoadComplete());
+    LSQueue->insert({lsq_hash, req});
+
+    // Reverted to original approach for type-specific reads
     if(nleft >= 8){
-      readfunc(reinterpret_cast<uint64_t*>(ECALL.buf.data()));
+      mem->ReadVal<uint64_t>(HartToExec, addr+ECALL.string.size(), reinterpret_cast<uint64_t*>(ECALL.buf.data()), req, REVMEM_FLAGS(0));
+      ECALL.bytesRead = 8;
     } else if(nleft >= 4){
-      readfunc(reinterpret_cast<uint32_t*>(ECALL.buf.data()));
+      mem->ReadVal<uint32_t>(HartToExec, addr+ECALL.string.size(), reinterpret_cast<uint32_t*>(ECALL.buf.data()), req, REVMEM_FLAGS(0));
```

Comment:
The explicit template argument `ReadVal<uint32_t>` is unnecessary. You can call `ReadVal()` without the template argument. It is automatically deduced from `reinterpret_cast<uint32_t*>(ECALL.buf.data())`. Same for all of the other cases.

Reply:
```c++
  template <typename T>
  bool ReadVal( unsigned Hart, uint64_t Addr, T *Target,
                const MemReq& req,
                StandardMem::Request::flags_t flags){
    return ReadMem(Hart, Addr, sizeof(T), Target, req, flags);
  }
```
The third `ReadVal` parameter is `T *Target`. Passing `reintrepret_cast<U*>(ECALL.buf.data())` as its third argument causes `T` to be automatically deduced to `U`. You do not need to pass `T` explicitly.

Reply:
Gotcha... I removed it